modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/index/sorted/PageMemorySortedIndexStorage.java

```diff
@@ -396,4 +372,56 @@ private int compareRows(SortedIndexRowKey key1, SortedIndexRowKey key2) {
             );
         }
     }
+
+    /**
+     * Prepares storage for rebalancing.
+     *
+     * <p>Stops ongoing index operations.
+     *
+     * @throws StorageRebalanceException If there was an error when starting the rebalance.
+     */
+    public void startRebalance() {
+        if (!state.compareAndSet(StorageState.RUNNABLE, StorageState.REBALANCE)) {
+            throwExceptionDependingOnStorageStateOnRebalance(state.get(), createStorageInfo());
+        }
+
+        // Stops ongoing operations on the storage.
```

Comment:
I don't think that it properly reflects what you're doing. Basically, you propagate that state that you just set to all further storage operations, while also waiting that all operations that saw the previous state are already completed.

Reply:
I tried to modify it.